Simulator/Utils/Factory.cpp

```diff
@@ -0,0 +1,120 @@
+
+// Layout
+#include "Layouts/NG911/Layout911.h"
+#include "Layouts/Neuro/DynamicLayout.h"
+#include "Layouts/Neuro/FixedLayout.h"
+
+// Vertices
+#include "Vertices/NG911/All911Vertices.h"
+#include "Vertices/Neuro/AllIZHNeurons.h"
+#include "Vertices/Neuro/AllLIFNeurons.h"
+
+// Recorder
+#include "Recorders/IRecorder.h"
+#include "Recorders/NG911/Xml911Recorder.h"
+#include "Recorders/Neuro/XmlGrowthRecorder.h"
+#include "Recorders/Neuro/XmlSTDPRecorder.h"
+#include "Recorders/XmlRecorder.h"
+
+#if defined(HDF5)
+   #include "Recorders/Hdf5Recorder.h"
+   #include "Recorders/Neuro/Hdf5GrowthRecorder.h"
+#endif
+
+//MTRand
+#include "RNG/MTRand.h"
+#include "RNG/Norm.h"
+
+template <typename T> std::map<std::string, T *(*)(void)> getCreateFunctionForType(T *placeholder)
+{
+   std::map<std::string, T *(*)(void)> createFunctionMap;
+
+   // Register Connections classes
+   if constexpr (std::is_same_v<T, Connections>) {
+      createFunctionMap["ConnStatic"] = &ConnStatic::Create;
+      createFunctionMap["ConnGrowth"] = &ConnGrowth::Create;
+      createFunctionMap["Connections911"] = &Connections911::Create;
+   }
+
+   // Register AllEdges classes
+   else if constexpr (std::is_same_v<T, AllEdges>) {
+      createFunctionMap["AllSpikingSynapses"] = &AllSpikingSynapses::Create;
+      createFunctionMap["AllSTDPSynapses"] = &AllSTDPSynapses::Create;
+      createFunctionMap["AllDSSynapses"] = &AllDSSynapses::Create;
+      createFunctionMap["AllDynamicSTDPSynapses"] = &AllDynamicSTDPSynapses::Create;
+      createFunctionMap["All911Edges"] = &All911Edges::Create;
+   }
+
+   // Register Layout classes
+   else if constexpr (std::is_same_v<T, Layout>) {
+      createFunctionMap["FixedLayout"] = &FixedLayout::Create;
+      createFunctionMap["DynamicLayout"] = &DynamicLayout::Create;
+      createFunctionMap["Layout911"] = &Layout911::Create;
+   }
+
+   // Register AllVertices classes
+   else if constexpr (std::is_same_v<T, AllVertices>) {
+      createFunctionMap["AllLIFNeurons"] = &AllLIFNeurons::Create;
+      createFunctionMap["AllIZHNeurons"] = &AllIZHNeurons::Create;
+      createFunctionMap["All911Vertices"] = &All911Vertices::Create;
+   }
+
+   // Register IRecorder classes
+   else if constexpr (std::is_same_v<T, IRecorder>) {
+      createFunctionMap["XmlRecorder"] = &XmlRecorder::Create;
+      createFunctionMap["XmlGrowthRecorder"] = &XmlGrowthRecorder::Create;
+      createFunctionMap["XmlSTDPRecorder"] = &XmlSTDPRecorder::Create;
+      createFunctionMap["Xml911Recorder"] = &Xml911Recorder::Create;
+#if defined(HDF5)
+      createFunctionMap["Hdf5Recorder"] = &Hdf5Recorder::Create;
+      createFunctionMap["Hdf5GrowthRecorder"] = &Hdf5GrowthRecorder::Create;
+#endif
+   }
+
+   // Register MTRand classes
+   else if constexpr (std::is_same_v<T, MTRand>) {
+      createFunctionMap["MTRand"] = &MTRand::Create;
+      createFunctionMap["Norm"] = &Norm::Create;
+   }
+   return createFunctionMap;
```

Comment:
Is there any way to guard against `T` not matching any of the tests? Or should the empty map be detected by the calling function? By the way, can this whole function be `constexpr`?

Reply:
Added a static_assert to check for unexpected types.

Reply:
>  can this whole function be constexpr?

Unfortunately, std::map is not constexpr compliant, however, we can create a template to generate a constexpr map with the help of std::arrays and std::pair, but it might add a lot of complexity to the existing code